src/class/builtin_classes/file_input_stream.rs

```diff
@@ -0,0 +1,111 @@
+use std::{
+    any::Any,
+    io::{self, Read},
+    rc::Rc,
+};
+
+use crate::{
+    class::{
+        ArgumentKind, Class, ClassInstance, Field, Method, MethodCode,
+        RustMethodReturn, SimpleArgumentKind,
+    },
+    executor::{frame_stack::StackValue, Frame},
+};
+
+pub struct FileInputStream {
+    methods: Vec<Rc<Method>>,
+}
+
+impl FileInputStream {
+    pub fn new() -> FileInputStream {
+        FileInputStream {
+            methods: vec![Rc::new(Method {
+                code: MethodCode::Rust(read),
+                name: "read".to_owned(),
+                parameters: vec![],
+                return_type: Some(ArgumentKind::Simple(
+                    SimpleArgumentKind::Int,
+                )),
+                is_static: false,
+            })],
+        }
+    }
+
+    pub fn new_instance(self: &Rc<Self>) -> FileInputStreamInstance {
+        FileInputStreamInstance {
+            class: self.clone(),
+        }
+    }
+}
+
+impl Default for FileInputStream {
+    fn default() -> Self {
+        FileInputStream::new()
+    }
+}
+
+fn read(frame: &mut Frame) -> RustMethodReturn {
+    let input = io::stdin()
+        .bytes()
+        .next()
+        .expect("some input on stdin")
+        .unwrap();
+    frame
+        .operand_stack
+        .push(StackValue::Int(input.into()))
+        .unwrap();
+    RustMethodReturn::Void
```

Comment:
correctness: I don't understand why you do not return the int here, but push it onto the stack. According to line 22 following this method should return int, doesn't it?

Reply:
Hmm, as I remember our thought behind the `RustMethodReturn` was to be able to return some internal meta information to the executor - not for the actual return values of Java methods.

Reply:
As far as I can rembember:

- we have the `Update` enum returned by `OpCode::execute` - this one can return meta information to the executor, e. g. for goto (which includes all the if variants we implemented last week)
- we have the `RustMethodReturn` enum - this one simply returns the return value of the function, I believe, for the following reasons:
  - if it were to carry any metadata, we surely wouldn't have named the empty variant after Java's `void` return type
  - what kind of meta information would it return? jumps are always local to the current function, which has ended at the point a `RustMethod` returns. (Exceptions would be the exception here (pun unintended), but I don't think we modeled with them in mind at any point so far)
  - If we would extract the return value from the stack, we:
    -  wouldn't initialize the stack for `RustMethod` with size 0 (https://github.com/cmjava2023/vm/blob/d0fccee98d16efef74741c668e35da2311d25551/src/executor.rs#L98)
    - probably would already have implemented extraction of the return value from the stack; instead we have a match statement which so far only matches the `RustMethodReturn::Void` variant - since we didn't implement any other return values so far (https://github.com/cmjava2023/vm/blob/d0fccee98d16efef74741c668e35da2311d25551/src/executor.rs#L109)

---

Though no matter on how exactly we define the meaning of the operand stack and the `RustMethodReturn` enum now, for this to work we will have to implement the actual 'return value' part in the executor, i.e. [inside the match statement that actually calls the `RustMethod`](https://github.com/cmjava2023/vm/blob/d0fccee98d16efef74741c668e35da2311d25551/src/executor.rs#L108)